### PR TITLE
add create, search and get medication order implementation

### PR DESCRIPTION
**Pull request by kwateng**

### CODE REVIEW

File: api/src/main/java/org/openmrs/module/fhir/api/MedicationRequestService.java [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+     */
+    MedicationRequest getMedicationRequestById(String uuid);
+
+    /**
+     * Search drug orders list by uuid
+     *
+     * @param uuid of the drug order
+     * @return drug order obj
+     */
+    List<MedicationRequest> searchAppointmentById(String uuid);
```
Comment (harsha89) [inline api/src/main/java/org/openmrs/module/fhir/api/MedicationRequestService.java]: Appointment should be renamed

File: api/src/main/java/org/openmrs/module/fhir/api/impl/MedicationRequestServiceImpl.java [hunk 2]
```diff
@@ -0,0 +1,118 @@
+package org.openmrs.module.fhir.api.impl;
```
Comment (harsha89) [inline api/src/main/java/org/openmrs/module/fhir/api/impl/MedicationRequestServiceImpl.java]: Licence missing

File: api/src/main/java/org/openmrs/module/fhir/api/impl/MedicationRequestServiceImpl.java [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+        if (drugOrder == null || drugOrder.isVoided()) {
+            return null;
+        }
+        return FHIRMedicationRequestUtil.generateMedicationRequest(drugOrder);
+    }
+
+    /**
+     * @see MedicationRequestService#searchAppointmentById(String)
+     */
+    public List<MedicationRequest> searchAppointmentById(String uuid) {
```
Comment (harsha89) [inline api/src/main/java/org/openmrs/module/fhir/api/impl/MedicationRequestServiceImpl.java]: appointment?

File: api/src/main/java/org/openmrs/module/fhir/api/impl/MedicationRequestServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+                StringBuilder errorMessage = new StringBuilder("The request cannot be processed due to the following issues \n");
+                for (int i = 0; i < errors.size(); i++) {
+                    errorMessage.append((i + 1) + " : " + errors.get(i) + "\n");
+                }
+                throw new UnprocessableEntityException(errorMessage.toString());
+            }
+            requestDrugOrder = (DrugOrder) Context.getOrderService().saveOrder(generatedDrugOrder, null);
+            return FHIRMedicationRequestUtil.generateMedicationRequest(requestDrugOrder);
+        } else {
+            if (medicationRequest.getId() == null) { // since we need to PUT the observation to a specific URI, we need to set the uuid
```
Comment (harsha89) [inline api/src/main/java/org/openmrs/module/fhir/api/impl/MedicationRequestServiceImpl.java]: correct the commnet

File: api/src/main/java/org/openmrs/module/fhir/api/impl/MedicationRequestServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+        drugOrder = (DrugOrder) Context.getOrderService().saveOrder(drugOrder, null);
+        return FHIRMedicationRequestUtil.generateMedicationRequest(drugOrder);
+    }
+
+    /**
+     * @see MedicationRequestService#updateFHIRMedicationRequest(MedicationRequest, String)
+     */
+    public MedicationRequest updateFHIRMedicationRequest(MedicationRequest medicationRequest, String uuid) {
+        List<String> errors = new ArrayList<String>();
+        DrugOrder requestDrugOrder = FHIRMedicationRequestUtil.generateDrugOrder(medicationRequest, errors);
```
Comment (harsha89) [inline api/src/main/java/org/openmrs/module/fhir/api/impl/MedicationRequestServiceImpl.java]: use better variable names

File: api/src/main/java/org/openmrs/module/fhir/api/util/FHIRMedicationRequestUtil.java [hunk 6]
```diff
@@ -0,0 +1,215 @@
...
+
+public class FHIRMedicationRequestUtil {
+
+    public static MedicationRequest generateMedicationRequest(org.openmrs.DrugOrder omrsDrugOrder) {
+        MedicationRequest request = new MedicationRequest();
+        if(!StringUtils.isEmpty(omrsDrugOrder.getUuid())) {
+            request.setId(omrsDrugOrder.getUuid());
+        }
+
+        Reference patientRef = FHIRPatientUtil.buildPatientReference(omrsDrugOrder.getPatient());
```
Comment (harsha89) [inline api/src/main/java/org/openmrs/module/fhir/api/util/FHIRMedicationRequestUtil.java]: add comments to mappings. It helps to understand more

File: api/src/main/java/org/openmrs/module/fhir/api/util/FHIRMedicationRequestUtil.java [hunk 7]
```diff
@@ -0,0 +1,215 @@
...
+            if (timingMappings != null && !timingMappings.isEmpty()) {
+                for (ConceptMap map : timingMappings) {
+                    if (map.getConceptReferenceTerm() != null) {
+                        timingDts.add(FHIRUtils.getCodingDtByConceptMappings(map));
+                    }
+                }
+            }
+
+            //Setting default omrs concept
+            if (orderFrequency.getConcept().getName() != null) {
```
Comment (harsha89) [inline api/src/main/java/org/openmrs/module/fhir/api/util/FHIRMedicationRequestUtil.java]: can't we use common method. Same thing repating

File: api/src/main/java/org/openmrs/module/fhir/api/util/FHIRMedicationRequestUtil.java [hunk 8]
```diff
@@ -0,0 +1,215 @@
...
+                    new MedicationRequest.MedicationRequestRequesterComponent();
+            reqComponent.setAgent(providerRef);
+            request.setRequester(reqComponent);
+            request.setRecorder(providerRef);
+        }
+        return request;
+    }
+
+    public static DrugOrder generateDrugOrder(MedicationRequest fhirMedicationRequest, List<String> errors) {
+        return null;
```
Comment (harsha89) [inline api/src/main/java/org/openmrs/module/fhir/api/util/FHIRMedicationRequestUtil.java]: add TODOs

File: api/src/main/java/org/openmrs/module/fhir/api/util/FHIRObsUtil.java [hunk 9]
```diff
@@ -422,5 +422,11 @@ public static Obs copyObsAttributes(Obs requestObs, Obs retrievedObs, List<Strin
 		retrievedObs.setComment(requestObs.getComment());
 		return retrievedObs;
 	}
-	
+
+	public static Reference getFHIREncounterReference(Encounter encounter) {
```
Comment (harsha89) [inline api/src/main/java/org/openmrs/module/fhir/api/util/FHIRObsUtil.java]: method comment is missing

File: api/src/main/java/org/openmrs/module/fhir/api/util/FHIRPatientUtil.java [hunk 10]
```diff
@@ -358,4 +359,24 @@ public static Patient generatePatient(org.openmrs.Patient omrsPatient) {
 		return retrievedPatient;
 	}
 
+	public static Reference buildPatientReference(org.openmrs.Patient patient) {
```
Comment (harsha89) [inline api/src/main/java/org/openmrs/module/fhir/api/util/FHIRPatientUtil.java]: method comment is missing

File: api/src/main/java/org/openmrs/module/fhir/api/util/FHIRPractitionerUtil.java [hunk 11]
```diff
@@ -411,4 +412,24 @@ public static Person extractOpenMRSPerson(Practitioner practitioner) {
 		return retrievedProvider;
 	}
 
+	public static Reference buildPractionaerReference(org.openmrs.Provider provider) {
```
Comment (harsha89) [inline api/src/main/java/org/openmrs/module/fhir/api/util/FHIRPractitionerUtil.java]: method comment is missing

File: api/src/test/java/org/openmrs/module/fhir/api/MedicationRequestTest.java [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+	@Test
+	public void getPractitioner_shouldReturnResourceIfExists() {
+		String medicationRequestUuid = "56b9196c-bcac-4c2f-b3a2-123464a96439";
+		MedicationRequest medicationRequest = getService().getMedicationRequestById(medicationRequestUuid);
+		assertNotNull(medicationRequest);
+		assertEquals(medicationRequestUuid, medicationRequest.getId().toString());
+	}
+
+	@Test
+	public void SearchPractitionerById_shouldReturnBundle() throws FHIRValidationException {
```
Comment (harsha89) [inline api/src/test/java/org/openmrs/module/fhir/api/MedicationRequestTest.java]: starts with simple. Practitioner? should be renamed

File: api/src/test/java/org/openmrs/module/fhir/api/MedicationRequestTest.java [hunk 13]
```diff
@@ -0,0 +1,82 @@
...
+		executeDataSet(DRUG_ORDER_INITIAL_DATA_XML);
+	}
+
+	@Test
+	public void shouldSetupContext() {
+		assertNotNull(getService());
+	}
+
+	@Test
+	public void getPractitioner_shouldReturnResourceIfExists() {
```
Comment (harsha89) [inline api/src/test/java/org/openmrs/module/fhir/api/MedicationRequestTest.java]: Practitioner? should be renamed

**Comment by harsha89** [comment]
Please add the creation of the medication request to this pull request. Use meaningful commit comments.

**Comment by harsha89** [comment]
@kwateng please create a jira and add this pull request.

**Comment by kwateng** [comment]
@harsha89 https://issues.openmrs.org/browse/FM-200 created

**Pull request merged**